packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -99,124 +107,330 @@ const defaultProps = {
+        if (type === DateUtil.inputType.MONTH) {
+          dayFocus.focus();
+        } else {
+          yearFocus.focus();
+        }
+      }
+    } else if (dateVariant === 'DD-MM-YYYY') {
+      if (inputValue.length === 2) {
+        if (type === DateUtil.inputType.DAY) {
+          monthFocus.focus();
+        } else {
+          yearFocus.focus();
+        }
+      }
+    } else if (dateVariant === 'YYYY-MM-DD') {
+      if (inputValue.length === 2 && type === DateUtil.inputType.MONTH) {
+        dayFocus.focus();
+      } else if (inputValue.length === 4) {
+        monthFocus.focus();
+      }
     }
-  }
+  };
 
-  handleOnChange(event) {
-    if (!DateUtil.validDateInput(event.target.value)) {
-      return;
+  /**
+   * Sets state for the day, month, and year.
+   * @param {string} inputValue - The input value to set in state.
+   * @param {int} type - The inputType (day, month, or year).
+   */
+  const setDate = (inputValue, type) => {
+    if (type === DateUtil.inputType.DAY) {
+      dateDispatch({ day: inputValue });
+    } else if (type === DateUtil.inputType.MONTH) {
+      dateDispatch({ month: inputValue });
+    } else {
+      dateDispatch({ year: inputValue });
     }
+    moveFocusOnChange(inputValue, type, localeFormat);
+  };
+
+  const handleDateChange = (event, type) => {
+    const inputValue = event.target.value.replace(/\D/gm, '');
 
-    if (this.props.onChange) {
-      this.props.onChange(event);
+    /**
+     * Sets the day, month and year based on input values, formats them
+     * based on the date format variant, and passes the formatted date to onChange.
+     */
+    const day = type === DateUtil.inputType.DAY ? inputValue : date.day;
+    const month = type === DateUtil.inputType.MONTH ? inputValue : date.month;
+    const year = type === DateUtil.inputType.YEAR ? inputValue : date.year;
+    let inputDate;
+    let formattedDate;
+    if (day.length === 2 && month.length === 2 && year.length === 4) {
+      inputDate = DateUtil.formatISODate(`${year}-${month}-${day}`, 'YYYY-MM-DD');
+      formattedDate = DateUtil.strictFormatISODate(inputDate, dateFormat);
     }
-  }
 
-  handleOnKeyDown(event) {
-    if (this.props.onKeyDown) {
-      this.props.onKeyDown(event);
+    /* eslint-disable no-param-reassign */
+    if (DateUtil.isValidDate(formattedDate, dateFormat)) {
+      event.target.value = formattedDate;
+      onChange(event);
+    } else if (day === '' && month === '' && year === '') {
+      onChange(event);
+    } else if (day.length < 2 || month.length < 2 || year.length < 4) {
+      // Handles raw input change updates by combining inputs.
+      event.preventDefault();
+      if (localeFormat === 'MM-DD-YYYY') {
+        event.target.value = month + day + year;
+      } else if (localeFormat === 'DD-MM-YYYY') {
+        event.target.value = day + month + year;
+      } else {
+        event.target.value = year + month + day;
+      }
+      onChange(event);
```

Comment:
- If I entered a complete date in the MM/DD/YYYY format but it is invalid (For example, `12/99/2020`) it would not satisfy any of the above statements and as a result, it would not trigger the `onChangeRaw` event. The `onChangeRaw` event would continue to get called for each entry I make up to the last digit because it gets in the 3rd `else if` but once I enter the last digit to complete a 4-digit year, it falls thru these checks so none of the `onChange` above gets hit.

- Previously the value returned in the onChangeRaw callback is whatever the user enters. For example, if they enter "01/01/2020" they would get back "01/01/2020. Now with the separator provided, they would just enter "01" followed by "01" followed by "2020". However, the value in onChangeRaw they now get is "01012020" without the separator. I think that is the right value to return since technically those are the only values they entered. However, from a consumer standpoint it would probably be hard for them to parse or get anything out of "01012020" if they need to convert it back to an ISO-8601 format. Should we inject the separators in the value to make easier for parsing?

I'm interested to hear what other people think.

![Screen Shot 2020-04-07 at 3 23 23 PM](https://user-images.githubusercontent.com/22618655/78715951-c2d6fc80-78e3-11ea-8da4-b34e4cb162db.png)